crates/amaru-consensus/src/consensus/validate_header.rs

```diff
@@ -220,19 +206,72 @@ impl ValidateHeader {
     }
 }
 
+#[derive(Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct ValidationFailed {
+    pub peer: Peer,
+    pub error: String,
+    pub action: UpstreamAction,
+}
+
+impl ValidationFailed {
+    pub fn kick_peer(peer: Peer, error: String) -> Self {
+        Self {
+            peer,
+            error,
+            action: UpstreamAction::KickPeer,
+        }
+    }
+
+    pub fn ban_peer(peer: Peer, error: String) -> Self {
+        Self {
+            peer,
+            error,
+            action: UpstreamAction::BanPeer,
+        }
+    }
+
+    pub fn only_logging(peer: Peer, error: String) -> Self {
+        Self {
+            peer,
+            error,
+            action: UpstreamAction::OnlyLogging,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+pub enum UpstreamAction {
+    KickPeer,
+    BanPeer,
+    OnlyLogging,
+}
+
 type State = (
     ValidateHeader,
     GlobalParameters,
     StageRef<DecodedChainSyncEvent, Void>,
+    StageRef<ValidationFailed, Void>,
 );
 
 pub async fn stage(
     state: State,
     msg: DecodedChainSyncEvent,
     eff: Effects<DecodedChainSyncEvent, State>,
-) -> Result<State, anyhow::Error> {
-    let (mut state, global, downstream) = state;
-    let result = state.validate_header(&eff, msg, &global).await?;
+) -> State {
+    let (mut state, global, downstream, errors) = state;
+    let peer = msg.peer();
+    let result = match state.validate_header(&eff, msg, &global).await {
+        Ok(result) => result,
+        Err(error) => {
+            tracing::warn!(%peer, %error, "invalid header");
+            eff.send(
+                &errors,
+                ValidationFailed::kick_peer(peer, error.to_string()),
```

Comment:
I don't think it's the job of `validate_header` stage to decide to `kick_peer`. The whole point of the `error_handling` stage is to take this decision and channel it to the right place.

Reply:
Yes, I did it this way because at that time I thought I could get away without making all those error types `SendData` (which requires PartialEq, Serialize, Deserialize). Now that I’ve added those, I’ll move the decision to the error stage (which should probably be renamed to upstream_errors).